Dev/ad7768

acusco opened the pull request:
This PR makes some driver fixes and adds support for AD7768-4 ADC.

- Added number of data-lines the ADC conversions data are output on
- Added optional hard reset GPIO
- Added crc attribute - enables/disables crc calculation
- Compute the sampling frequency range according to mclk freq

tachicialex commented on drivers/iio/adc/ad7768.c:
```diff
@@ -639,7 +785,36 @@ static int ad7768_probe(struct spi_device *spi)
+					"failed to get the gpio_shutdown_n\n");
+	}
+
+	ret = of_property_read_u32(st->spi->dev.of_node, "adi,data-lines",
```
Throw an error if st->datalines == 0 here.

tachicialex commented on drivers/iio/adc/ad7768.c:
```diff
@@ -639,7 +785,36 @@ static int ad7768_probe(struct spi_device *spi)
 
-	ret = ad7768_set_sampling_freq(indio_dev, AD7768_MAX_SAMP_FREQ);
+	/* get out of reset state */
+	st->gpio_reset_n = devm_gpiod_get_optional(&spi->dev, "reset-n",
```
The reset part is a bit weird.  Maybe make sure this GPIO goes low via gpiod_set_value().
Do smth like:
gpiod_set_value(1)
sleep
gpiod_set_value(0)
Also datasheet specifies a minimum amount of time to stay in this state.

tachicialex commented on Documentation/devicetree/bindings/iio/adc/adi,ad7768.txt:
```diff
@@ -1,14 +1,15 @@
 	- clocks: phandle to the master clock (mclk).
 	- clock-names: Must be "mclk".
+	- adi,data-lines: Must be 1/2/8 for AD7768 or 1/4 for AD7768-4 ADC.
 
```
Document here the reset-gpio shutdown property and the fact that they need to be active_low. Polarity is not always obvious.

nunojsa replied:
Just to make it clear. That should be on a different patch

tachicialex commented on drivers/iio/adc/ad7768.c:
```diff
@@ -217,28 +225,33 @@ static int ad7768_set_clk_divs(struct ad7768_state *st,
 		return -EINVAL;
 
+	dclk = result * SAMPLE_SIZE * chan_per_doutx;
+	dclk_div = DIV_ROUND_CLOSEST_ULL(mclk, dclk);
```
error on 0, with a message that something went wrong?

tachicialex reviewed (changes requested):
looks good

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -158,7 +163,7 @@ static int ad7768_spi_reg_write(struct ad7768_state *st,
 static int ad7768_spi_write_mask(struct ad7768_state *st,
 				 unsigned int addr,
-				 unsigned long int mask,
+				 unsigned long mask,
```
why this change? At first glance it seems unrelated to this patch

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -66,6 +66,7 @@
 
 #define SAMPLE_SIZE				32
 #define NUM_CHANNELS				8
+
```
unrelated change

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -54,6 +54,11 @@
+#define AD7768_INTERFACE_CFG_CRC_SELECT_MSK	GENMASK(3, 2)
+/* only 4 samples CRC calculation support exists */
+#define AD7768_INTERFACE_CFG_CRC_SELECT(x)	(((x) & 0x3) << 2)
+#define AD7768_INTERFACE_CFG_GET_CRC_STATUS(x)	(((x) >> 2) & 0x3)
```
I guess FIELD_PREP() and FIELD_GET() can be used... but if that is not being used in the rest of the file, better to keep consistency

pcercuei replied:
Even better, convert the file to use FIELD_PREP / FIELD_GET before this commit ;)

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -750,8 +792,14 @@ static int ad7768_probe(struct spi_device *spi)
 
-	max_samp_freq = (st->datalines == 1) ? AD7768_MAX_SAMP_FREQ_1LINE
-					    : AD7768_MAX_SAMP_FREQ_MLINES;
+	st->power_modes_info = ad7768_power_modes_info;
```
I think this will fail if you have multiple instances of this device as you're  only assigning the pointer (and ad7768_power_modes_info is static). You need to copy the complete structure. So just make `st->power_modes_info` as a variable and then copy the whole thing. Or since we just have those 3 mclk_div values, maybe just drop the static structure and do the assignments in `ad7768_set_power_modes_info()`. I guess it will be cheaper than the copy but does not really matter as this is not a fast path

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -471,23 +490,46 @@ static int ad7768_set_sampling_freq(struct iio_dev *dev,
+			if (st->power_modes_info[mode].mclk_div != 0) {
+				sampl_freq = mclk / (ad7768_dec_rate[dec]
+					* st->power_modes_info[mode].mclk_div);
+			}
```
This is very unreadable. So, I would either use helper variables for st->power_modes_info[mode].mclk_div and the decimation rate or just use an helper function to deal with all of the inner code...

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -787,6 +792,14 @@ static int ad7768_probe(struct spi_device *spi)
+	if (axiadc_chip_info->num_channels == 4) {
+		st->gpio_shutdown_n = devm_gpiod_get_optional(&spi->dev,
+					"shutdown-n", GPIOD_OUT_LOW);
+	if (IS_ERR(st->gpio_shutdown_n))
```
indentation...

amiclaus commented on drivers/iio/adc/ad7768.c:
```diff
@@ -114,7 +109,6 @@ static const int ad7768_dec_rate[6] = {
 	32, 64, 128, 256, 512, 1024
 };
 
-
```
nit: unrelated change

amiclaus commented on drivers/iio/adc/ad7768.c:
```diff
@@ -283,14 +295,15 @@ static int ad7768_set_power_mode(struct iio_dev *dev,
 				    regval);
 	if (ret < 0)
 		return ret;
+
```
nit: unrelated change

amiclaus commented on drivers/iio/adc/ad7768.c:
```diff
@@ -441,21 +454,27 @@ static int ad7768_set_sampling_freq(struct iio_dev *dev,
+		power_mode = st->power_mode;
 	}
 
+
```
nit: unrelated change

amiclaus reviewed (commented):
just some nits on my side, besides the rest of the comments

pcercuei commented on drivers/iio/adc/cf_axi_adc_core.c:
```diff
@@ -994,6 +994,7 @@ static const struct of_device_id axiadc_of_match[] = {
 	{ .compatible = "adi,axi-adc-10.0.a", .data = &axi_adc_10_0_a_info },
 	{ .compatible = "adi,axi-adrv9002-rx-1.0", .data = &axi_adc_10_1_b_info},
 	{ .compatible = "adi,axi-ad9083-rx-1.0", .data = &axi_adc_10_0_a_info },
+	{ .compatible = "adi,axi-ad7768-rx-1.0", .data = &axi_adc_10_0_a_info },
```
Shouldn't it simply use `adi,axi-adc-10.0.a` as the fallback string? Then the driver doesn't need to be updated.

nunojsa replied:
yeah, that's right...

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -51,9 +51,11 @@
 /* AD7768_INTERFACE_CFG */
 #define AD7768_INTERFACE_CFG_DCLK_DIV_MSK	GENMASK(1, 0)
-#define AD7768_INTERFACE_CFG_DCLK_DIV_MODE(x)	(((x) & 0x3) << 0)
+#define AD7768_INTERFACE_CFG_DCLK_DIV_MODE(x)	(((3-ilog2(x)) & 0x3) << 0)
```
The commit doesn't touch this line though.

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -757,6 +756,12 @@ static int ad7768_probe(struct spi_device *spi)
+	axiadc_chip_info = device_get_match_data(&spi->dev);
+	if (!axiadc_chip_info)
+		axiadc_chip_info =
+			(const struct axiadc_chip_info *)spi_get_device_id(spi)->driver_data;
```
It should already be a `void *`, so you don't need to cast it again.
`chip_info = spi_get_device_id(spi)->driver_data;`

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -657,14 +832,23 @@ static int ad7768_probe(struct spi_device *spi)
 static const struct spi_device_id ad7768_id[] = {
-	{"ad7768", ID_AD7768},
+	{"ad7768", ID_AD7768, (kernel_ulong_t)&conv_chip_info_tbl[ID_AD7768]},
+	{"ad7768-4", ID_AD7768_4, (kernel_ulong_t)&conv_chip_info_tbl[ID_AD7768_4]},
```
And how do you cram 3 parameters into a structure that only has two fields? ...

bbushey commented:
Hi All,

Looks like this merge hasn't happened yet, can you please finish this review so that we can have the merge occur?  This product is used in several upcoming designs, and the driver updates are a bottleneck at this point.

acusco commented:
V1 -> V2

- fixed checkpatch errors
- fixed nits
- rewritten reset part
- edited too long commit messages
- removed crc status attribute (enabled crc all the time)
- refactored unreadable code
- removed static struct ad7768_power_modes_info and did all the assignments in _set_power_modes_info
- removed shutdown_gpio
- restricted data-lines property to specific values

pcercuei commented:
Note that if this driver is headed upstream, maintainers will most certainly complain about the documentation being a .txt instead of a YAML schema.

nunojsa commented:
> Note that if this driver is headed upstream, maintainers will most certainly complain about the documentation being a .txt instead of a YAML schema.

Indeed... However, to mainline this thing we will need proper usptream support for `#include "cf_axi_adc.h"` (which I want to get started til the end of the month).

Nevertheless, we could have a patch to do the conversion to YAML already (as we are adding new properties)

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -642,7 +672,15 @@ static int ad7768_probe(struct spi_device *spi)
+	ret = of_property_read_u32(st->spi->dev.of_node, "adi,data-lines",
+				  &st->datalines);
+	if (ret < 0)
+		return ret;
```
Is `adi,data-lines` really a mandatory property for this part? Can't we default to some value?

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -676,6 +678,17 @@ static int ad7768_probe(struct spi_device *spi)
 
+	/* get out of reset state */
+	gpio_reset = devm_gpiod_get_optional(&spi->dev, "reset-n",
+						   GPIOD_OUT_HIGH);
```
nit... align `GPIOD_OUT_HIGH` with open parenthesis.

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -676,6 +678,17 @@ static int ad7768_probe(struct spi_device *spi)
+						   GPIOD_OUT_HIGH);
+	if (IS_ERR(gpio_reset))
+		return dev_err_probe(&spi->dev, PTR_ERR(gpio_reset),
+				     "failed to get the gpio reset\n");
```
nit: I would argue with `dev_err_probe()` just because the rest of the driver is not using it... So, this could be a full conversion patch. Not a big deal to me though...

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -676,6 +678,17 @@ static int ad7768_probe(struct spi_device *spi)
+		return dev_err_probe(&spi->dev, PTR_ERR(gpio_reset),
+				     "failed to get the gpio reset\n");
+	if (gpio_reset) {
+		usleep_range(1580, 1660);
```
could use `fsleep()`...

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -676,6 +678,17 @@ static int ad7768_probe(struct spi_device *spi)
+				     "failed to get the gpio reset\n");
+	if (gpio_reset) {
+		usleep_range(1580, 1660);
+		gpiod_set_value_cansleep(gpio_reset, 0);
```
shouldn't we also wait some time after bringing the device out of reset? Does the datasheet not specify any timing requirement?

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -223,28 +223,35 @@ static int ad7768_set_clk_divs(struct ad7768_state *st,
+	dclk_div = DIV_ROUND_CLOSEST_ULL(mclk, dclk);
+
+	if (dclk_div > AD7768_MAX_DCLK_DIV)
+		dclk_div = AD7768_MAX_DCLK_DIV;
```
Just one question: Is this change fixing something?

acusco replied:
> Just one question: Is this change fixing something?
There are only 4 possible values for dclk_div: 1, 2, 4, 8. If a user selects a configuration which is not valid it may happen  dclk_div to exceed max value.

acusco replied:
The frequencies will match because this does not influence the sampling frequency, `dclk_div` is used to set the `dclk` rate

nunojsa replied:
Hmm?? Am I missing something obvious in here?! 

So, `dclk_div` is given by `DIV_ROUND_CLOSEST_ULL(mclk, dclk);` and `dclk` is equal to `result * SAMPLE_SIZE * chan_per_doutx;`. AFAIC, `result == freq` and `freq` will be assigned to `st->sampling_freq` outside of this function, which means there's a clear relation between `dclk_div` and `freq`.

Also, what @pcercuei states in here:

> What guarantees you that you cannot have any other value for ddclk_div?

I'm kind of wondering the same? What guarantees you that `dclk_div` is not 5 for example? Maybe you should be using something like `roundup()` or `roundown()` and properly update the read side of the samplig frequency... If I'm not missing nothing:

`st->sampling_freq = mclk / (dclk_div * SAMPLE_SIZE * chan_per_doutx)`

acusco replied:
`st->sampling_freq = mclk / (mclk_div * dec_rate)`, the sampling frequency will always correspond to the value that user wants to write, `dclk_div` is used fot setting the division factor for dclk:  `dclk = mclk/dclk_div`

nunojsa replied:
So, you are telling me this: `st->sampling_freq = mclk / (mclk_div * dec_rate)` and then `dclk = mclk / dclk_div` which means `mclk = dclk * dclk_div` right? Now replace `mclk` and we have:

 `st->sampling_freq = dclk * dclk_div / (mclk_div * dec_rate)`

How can the samling_freq not depend on `dclk_div`?

It looks to me there's more into this.... Because on one side you have:

`mclk = freq * SAMPLE_SIZE * chan_per_doutx * dclk_div` 

and on the other side:

`mclk = freq * m_div * dec`

from the above you can tell that = `m_div * dec = SAMPLE_SIZE * chan_per_doutx * dclk_div` but it seems that's not always possible to achieve given the `chan_per_doutx` we have (which I guess it depends on hdl?). What you're doing is that if:

```
if (dclk_div > AD7768_MAX_DCLK_DIV)
	dclk_div = AD7768_MAX_DCLK_DIV;
```

Then you use the maximum allowed of 8 but this effectively means that DCLK will run faster than it should. And I have no idea how will the part behave  as your conversion path (which is affected by `m_div` + `dec_rate`) differs from your DCLK. As DCLK is running faster, I would say that it's likely your samples won't be valid.

I also resigned myself and just went to the datasheet and this looks very important:

```
Because there are eight channels and 32 bits of data per conversion, the conversion time and the setting of DCLK
directly determine the number of data output lines that are required via the FORMAT0 pin and FORMAT1 pin
settings on the AD7768, or the FORMAT0 pin on the AD7768-4. Therefore, the intended minimum decimation and
desired DCLK_DIV setting must be understood prior to choosing the setting of the FORMATx pins.
```

To me it looks like the available frequencies for each mode are not always possible as they depend on the number of data lines we are using.  So for the time being I would probably return error in this case (which is odd as we have an available_frequencies attr) but I guess we can see how viable it is to compute the possible frequencies given the number of lines we have...

But hey, I might completely off in here but I would surely have this understood and maybe ask the apps guy responsible for this part how this actually works.

pcercuei replied:
It doesn't cost much to add a
```
if (WARN_ON(hweight32(dclk_div) != 1)
    return -EINVAL;
```

nunojsa replied:
Well, I would not use `WARN_ON()` and IIUC, `hweight32()` won't be enough because `dclk_div` should always be a powe of 2. The problem is that it cannot be > 8

pcercuei replied:
Well, you say that `dclk_div` should always be a power of two, and I'm not convinced by that.

You are using DIV_ROUND_CLOSEST to divide an unknown value (mclk) using a divider obtained using another DIV_ROUND_CLOSEST, and you tell me you will always end up on a power of two?

Even if it was mathematically sound, which I doubt, I shouldn't have to do mental gymnastics to proof-read the sanity of a driver, where a simple value check will do.

nunojsa replied:
Actually, I was totally wrong because I was assuming these are power of 2 (for example assuming median mode):

`[AD7768_MEDIAN_MODE] = { 4000, 8000, 16000, 32000, 64000, 128000 },` 

Obviously they are not a power of 2 which means there's  no guarantee that `dclk` and `mclk` are... I was tricked by looking at this divided by 1000 (for some reason). So yeah, having something like this might make sense:

```
if  (hweight32(dclk_div) != 1 && dclk_div > AD7768_MAX_DCLK_DIV)
    return -EINVAL;
```

But this should be temporary because having an `ad7768_sampl_freq_avail` array, and then letting the user select one of them just to have an -EINVAL returned is very questionable :).

I'm convinced that given the dt property `adi,data-lines` we can compute the valid frequencies we have for each of the power modes and for the given `mclk` frequency...

This also means that we should cache the `mclk` rate otherwise we need to register a notifier in order to recompute the whole thing if the rate changes. IMO, that is overly complicating things as I doubt `mclk` will ever change on a real system

acusco replied:
Alright, I already compute the sampling frequencies dynamically but do not take into account the datalines no. I will rewrite the code following your suggestions

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -713,13 +711,24 @@ static int ad7768_probe(struct spi_device *spi)
+	st->chip_info = device_get_match_data(&spi->dev);
+	if (!st->chip_info) {
+		st->chip_info = (void *)spi_get_device_id(spi)->driver_data;
+
```
nit: no need for new line here

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -104,6 +109,25 @@ static const int ad7768_dec_rate[6] = {
+	},
+	[ID_AD7768_4] = {
+			 .len = ARRAY_SIZE(ad7768_4_available_datalines),
+			 .values = ad7768_4_available_datalines,
```
indentation looks odd

nunojsa replied:
I think it would be neater to have these tables directly in the chip_info structure. So, adding `struct ad7768_datalines_info` field into it. Or just the array and the len variable...

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -765,10 +802,13 @@ static int ad7768_probe(struct spi_device *spi)
-	if (ret < 0)
-		return ret;
-	if (!st->datalines)
-		return -EINVAL;
```
this is changing behavior... now the property is no longer mandatory which is not fine

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -765,10 +802,13 @@ static int ad7768_probe(struct spi_device *spi)
+					       ad7768_datalines_tbl[id].len,
+					       st->datalines);
+		if (ret < 0)
+			return ret;
```
well, this should be done from the beginning (in the patch that introduces this property) If easier, you might re-arrenge things, so you add the new device before introducing this property.

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -427,23 +450,53 @@ static int ad7768_set_sampling_freq(struct iio_dev *dev,
+	unsigned int mclk_div;
+	unsigned int dec_rate;
+	int max = 0, i = 0;
+
```
nit: no new line

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -397,21 +412,29 @@ static int ad7768_set_sampling_freq(struct iio_dev *dev,
+	mode_info = st->power_modes_info[st->power_mode];
+
+	if (freq < mode_info.min_freq || freq > mode_info.max_freq) {
+
```
nit: no new line

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -397,21 +412,29 @@ static int ad7768_set_sampling_freq(struct iio_dev *dev,
+	if (freq < mode_info.min_freq || freq > mode_info.max_freq) {
+
+		for (mode = 0; mode < AD7768_NUM_CONFIGS; mode++)
+
```
ditto

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -397,21 +412,29 @@ static int ad7768_set_sampling_freq(struct iio_dev *dev,
+		for (mode = 0; mode < AD7768_NUM_CONFIGS; mode++)
+
+			if (freq >= st->power_modes_info[mode].min_freq &&
+			    freq <= st->power_modes_info[mode].max_freq){
```
you need a space before `{`

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -264,6 +255,29 @@ static int ad7768_set_clk_divs(struct ad7768_state *st,
+	st->power_modes_info[AD7768_MEDIAN_MODE].mclk_div = 8;
+	st->power_modes_info[AD7768_FAST_MODE].mclk_div = 4;
+
+	for (i = 0; i < AD7768_NUM_CONFIGS; i++) {
```
nit: I would use `ARRAY_SIZE()`

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -264,6 +255,29 @@ static int ad7768_set_clk_divs(struct ad7768_state *st,
+
+		mode_info = &st->power_modes_info[i];
+
+		if (!mode_info->mclk_div)
```
Hmm is this actually possibly? Does not look like it is...

acusco replied:
Yes, it is, as the array size of power_modes_info if 4, but I only initialize structures corresponding to the power modes (0,2,3 array indices)

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -427,23 +450,53 @@ static int ad7768_set_sampling_freq(struct iio_dev *dev,
+		for (dec = AD7768_CONFIGS_PER_MODE - 1; dec >= 0;  dec--) {
+
+			if (!st->power_modes_info[mode].mclk_div)
+				continue;
```
don't think the above condition can ever happen (as you initialize everything at probe)

pcercuei replied:
If this check is justified (is it?) then it doesn't have to be nested 2 levels down, move it up one level (before the "for dec" loop).

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -713,13 +711,24 @@ static int ad7768_probe(struct spi_device *spi)
+
+		if (!st->chip_info)
+			return -EINVAL;
+	}
```
Please go with option 1... Ideally, I'd like to backport only bugfixes. For other things, I would like to have a strong reason for it and there are a lot of useful APIs that we still don't have so that this one is not special.

I know this is a simple addition and would not (likely) give me much trouble when merging to a kernel supporting it but still... I also know that then we end up with a lot of drivers that diverge from the mainline version but that's the price of being in older kernels...

pcercuei replied:
It was not a list of options ;)
If you don't want to cherry-pick the commit, then just remove the call to `device_get_match_data` (and the .data fields in ad7768_of_match) and get the pointer only from `spi_get_device_id(spi)->driver_data`.

nunojsa replied:
That's fine with me....

acusco commented on drivers/iio/adc/ad7768.c:
```diff
@@ -691,8 +689,8 @@ static int ad7768_register(struct ad7768_state *st, struct iio_dev *indio_dev)
-	indio_dev->channels = ad7768_channels;
-	indio_dev->num_channels = ARRAY_SIZE(ad7768_channels);
+	indio_dev->channels = st->chip_info->channel;
+	indio_dev->num_channels = st->chip_info->num_channels;
```
> You can keep the `ad7768_channels` with the 8 channels, and just have `num_channels` in your axiadc_chip_info structure.
> 
> If you keep `ad7768_channels` then maybe declare it like a regular array, not with a macro, because I have no idea why `DECLARE_AD7768_CHANNELS` was here in the first place.

Should I keep the channels info twice? This is already defined in axiadc_chip_info table which I keep in the chip state, and even if an axi adc is not present on the bus this info can be reused in ad7768_register. Correct me if I'm wrong.

acusco replied:
Should I keep the channels info twice? This is already defined in axiadc_chip_info table which I keep in the chip state, and even if an axi adc is not present on the bus this info can be reused in ad7768_register. Correct me if I'm wrong.

pcercuei replied:
I don't think you need those in the chip state, as you don't actually modify them; therefore they can be "static const" data.

bbushey commented:
Hi Paul and Nuno,

Is there an update on getting this merged to master?  This commit has some potential implications on releasing the next version of Kuiper (2021_R2) and we'd like to get this included if its a quick job

nunojsa commented:
> Hi Paul and Nuno,

> Is there an update on getting this merged to master? This commit has some potential implications on releasing the next version of Kuiper (2021_R2) and we'd like to get this included if its a quick job

Hi Brandon,

Well, that does not depend on me or @pcercuei. We did requested changes on this PR but AFAICT, nothing has changed so we cant approve it...

nunojsa commented:
> Hi Nuno, sorry for not understanding the full process here but who's responsible for updating the code based on your comments and then who is the one that can approve this? Seems to be sitting here for quite some time.

Updating the code should be done by the one opening the PR in the first place... Then, it has to be approved at least by one member of the linux team before I (or someone else) can merge it.

acusco commented:
V2 -> V3
The new version includes the requested changes.

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -272,6 +263,27 @@ static int ad7768_set_clk_divs(struct ad7768_state *st,
+			continue;
+
+		mode_info->min_freq = mclk / (mode_info->mclk_div * DEC_1024);
+		mode_info->max_freq = mclk / (mode_info->mclk_div * DEC_32);
```
If you really want to use macros, use `SZ_32` and `SZ_1K` from <linux/sizes.h>.

acusco replied:
The fact that `dlck_div` does only depend on `mclk_div`, `dec_rate` and `chan_per_doutx` which values are fixed and multiple of 2 (as `mclk` simplifies out of this equation), so `dclk_div` will also be a multiple of 2, but the max value can be 8 and that's why I have this constraint

acusco replied:
The fact that `dlck_div` does only depend on `mclk_div`, `dec_rate` and `chan_per_doutx` which values are fixed and multiple of 2 (as `mclk` simplifies out of this equation), so `dclk_div` will also be a multiple of 2, but the max value can be 8 and that's why I have this constraint

pcercuei replied:
`mclk` would simplify out of the equation if you worked with real numbers, not integers that are rounded up/down...

pcercuei reviewed (commented):
LGTM, a few things here and there that I would change, but overall it looks good.
And good job on the DT schema.

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ad7768.yaml:
```diff
@@ -0,0 +1,118 @@
+            adi,data-lines = <1>
+        }
+    };
+...
```
This is more a nit but I would do the conversion to yaml as the first commit. There's no point in adding new properties to the .txt file just to convert them later (in the same series) to yaml...

Also, are you removing the old .txt file?

pcercuei replied:
@acuso you can squash all the commits  that touch the YAML file into the commit that introduced it.

nunojsa reviewed (commented):
Looks almost there.. You still need, at least, to address CI complains:

https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=22169&view=logs&j=8816ca86-faec-5cfe-601f-19d1b3f9ab6b&t=2446816d-e5a7-5cd4-afcb-81c090218db4


https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=22169&view=logs&j=250c57b8-394d-549e-b22b-b888ca92eef2&t=376ad7af-102b-5c8d-9bc6-bdbbc87e4bef&l=961

acusco commented:
I pushed a new version with the last requested changes, but I still don't succeed to make the checkpatch happy :)

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -565,11 +578,43 @@ static void ad7768_clk_disable(void *data)
+	unsigned int i, len;
+
+	st->datalines = 1;
+	device_property_read_u32(&st->spi->dev, "adi,data-lines", &st->datalines);
```
nit: you could check the error code in here. If the property is not available you pretty much already now what to return right?

acusco replied:
why should I check the return status, if `st->datalines` is set to a default value before calling the property read function and if it is an optional property?

nunojsa replied:
Because if you don't have any property and you already set it to 1, you know it's valid right? So there's no need to run the following code to validate the value...

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -565,11 +578,43 @@ static void ad7768_clk_disable(void *data)
+
+	for (i = 0; i < len; i++) {
+		if (available_datalines[i] == st->datalines)
+			return i;
```
nit: `return 0`? Unless I'm missing something return > 0 has no special meaning for the caller...

pcercuei commented:
@acuso some of the checkpatch warnings can be addressed, though. Like the ones about the commit description's lines length and the one about the missing blank line.

nunojsa reviewed (commented):
Mostly nits. The only open question to me is the sampling frequency handling...

nunojsa commented:
@acusco, you can also run checkpatch locally so you can fix the issues before pushing (when CI starts to be annoying :D). Assuming your branch is properly rebased with master you can run `scripts/checkpatch.pl --strict --git origin/master..`

dbogdan commented:
@acusco, please remove the "Draft" tag when you think the changes are ready to be merged.

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -471,23 +490,46 @@ static int ad7768_set_sampling_freq(struct iio_dev *dev,
 	struct iio_dev *indio_dev = dev_to_iio_dev(dev);
 	struct ad7768_state *st;
 	int i, len = 0, size;
+	unsigned int sampl_freq_avail[9];
```
Yes, this looks fishy. The loop below will read the 9 values, the array is not zero-initialized and I don't see what guarantees that `ad7768_get_available_sampl_freq` will write exactly 9 values no less, no more.

acusco replied:
There are only 6 possible values for decimation rate and 3 power modes, so I think that the available sampling frequencies will always be 9 as they were defined in the old version, @pcercuei - if you have a suggestion of how can I do this differently, I'm willing to consider it

pcercuei replied:
The safer variant would be:
```
unsigned int sampl_freq_avail[AD7768_CONFIGS_PER_MODE * AD7768_NUM_CONFIGS] = { 0 };
```

pcercuei commented:
I reviewed the checkpatch CI issues and they can be ignored. It only fails because one commit description has a URL that's > 75 characters long.

dbogdan commented:
@pcercuei , I think the original changes you have requested were made. If this is the case, can you update (not necessarily approve the PR) your review? :)

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -565,11 +578,47 @@ static void ad7768_clk_disable(void *data)
+	ret = device_property_read_u32(&st->spi->dev, "adi,data-lines",
+				       &st->datalines);
+	if (ret < 0)
+		return 0;
```
nit: only return 0 if you get -EINVAL, and return "ret" if you get another error code.

pcercuei commented on Documentation/devicetree/bindings/iio/adc/adi,ad7768.yaml:
```diff
@@ -48,6 +48,18 @@ properties:
+      1 - one data-line (one DOUTx pin)
+      2 - two data-lines
+      4 - four data-lines
+      8 - eight data-lines
```
I think you can keep only the first line.

pcercuei commented:
I wouldn't ACK it yet, it is indubitably getting better but there are still some weird things in the code, and some of @nunojsa's comments were not addressed.

nunojsa commented:
> I wouldn't ACK it yet, it is indubitably getting better but there are still some weird things in the code, and some of @nunojsa's comments were not addressed.

I also had a discussion with @acusco yesterday which was (I believe) after the latest update. @acusco, is this PR already contemplating what we discussed yesterday?

To give some background, it appears that the chip will send zeroed samples (and marked as invalid) when dclk > samling_rate (of course only for the outstanding samples) but hdl is not discarding the invalid samples and we were not sure if they can do it or not.

Doing this in userspace is far from ideal so the only thing left is the original plan of making sure all the available sampling_frequencies can have a matched dclk. As it turns out, this might mean (depending on the number of data lines and mclk selected) that not all power modes will be available (it can happen that all the frequencies available for one mode are "invalid") which means the driver also has to take care about controlling which power modes are available or not...

acusco commented:
> All the sampling frequencies and power modes are available in this version, hdl is trying to discard zeroed samples

nunojsa commented:
> All the sampling frequencies and power modes are available in this version, hdl is trying to discard zeroed samples

Alright, that would make our life easier...

acusco commented:
V3 -> V4

- squash yaml commits
- check return value of datalines property
- refactor available_sampl_freq code
The problem with zeroed samples was solved by hdl, the driver supports all power modes and entire range of sampling frequencies for each of them.

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -232,28 +235,36 @@ static int ad7768_set_clk_divs(struct ad7768_state *st,
+
+	dclk_div = DIV_ROUND_CLOSEST_ULL(mclk, dclk);
+	if  (hweight32(dclk_div) != 1)
+		return -EINVAL;
```
This one is tricky... I would not treat this as an error, the same way we don't do with `dclk_div > AD7768_MAX_DCLK_DIV`.  As I see it, I would do something like:

```
if (dclk_div > AD7768_MAX_DCLK_DIV))
    dclk_div = AD7768_MAX_DCLK_DIV;
else if (hweight32(dclk_div) != 1)
    //round dclk_div to the next valid value...
```

You only have 4 possible values for this case, so you can even have some static table or something else more clever to choose the next valid value. We just need to keep in mind that we can't end up with dclk < sampling_rate * SAMPLE_SIZE * chan_per_doutx. Most likely, the safest thing to do (and maybe the only?) is to round down to the next valid value...

The reason why I'm complaining with returning error is because you give a list of available_frequencies to the user that are supposed to be valid. But then we return an error back in handling the data clock. This also brings another question to me: Is this condition ever possible?

```
if (freq != result)
     return -EINVAL;
```

Correct me if I'm wrong, I think in this precise patch, it's still possible to get the above condition but after your next patch (about dynamically calculating the available_freqs) this condition is no longer possible, right? I mean, unless there's a bug in your code :)

So I would either completely remove it or add some log before returning. Because, hitting that condition means there's a bug in the driver...

nunojsa reviewed (commented):
Alright... almost there! Just one (big) comment from my side.

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -41,6 +41,8 @@
 #define AD7768_POWER_MODE_MCLK_DIV_MSK		GENMASK(1, 0)
 #define AD7768_POWER_MODE_MCLK_DIV_MODE(x)	(((x) & 0x3) << 0)
+#define map_power_mode_to_regval(x)		((x > 0) ? x + 1 : x)
+#define map_regval_to_power_mode(x)		((x > 0) ? x - 1 : x)
```
nit: Typically this defines should have prefixed with the driver name (ad7768)

pcercuei replied:
This is confusing, I'm not sure that works.
`B = map_power_mode_to_regval(map_regval_to_power_mode(A))`
If you start with A == 1, you end up with B == 0.

nunojsa reviewed (approved):
Just some nits from my side...

@pcercuei , you were also very active in all this discussion. Could you also have a look to see if I missed something important?

pcercuei commented on arch/arm/boot/dts/socfpga_cyclone5_de10_nano_cn0579_i2c.dts:
```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
```
Just a nit on the licensing, is that one correct? The other .dts only use `GPL-2.0`.

acusco replied:
it's okay, all .dts related to this board use this licence

pcercuei commented on arch/arm/boot/dts/socfpga_cyclone5_de10_nano_cn0579_i2c.dts:
```diff
@@ -0,0 +1,95 @@
+};
+
+&fpga_axi {
+	rx_dma: rx-dmac@0x00028000 {
```
Try to use one of the node names recommended in the Device Tree specification, if possible (https://www.devicetree.org/specifications/). Also, the `0x` prefix should not appear when using an address in the node name.

Following the spec, this node should be called `dma-controller@00028000`.

This applies to a lot of nodes that you added across files, not just this instance here.

nunojsa replied:
> Following the spec, this node should be called dma-controller@00028000.

cool, didn't knew about that :+1:

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -82,34 +95,24 @@ struct ad7768_state {
 	ID_AD7768_4
 };
 
+
```
Unstage that empty line.

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -372,47 +405,29 @@ static const struct iio_enum ad7768_filter_type_iio_enum = {
-		return -EINVAL;
+	for (mode = 0; mode < AD7768_NUM_POWER_MODES; mode++) {
+		avail_freq = &st->avail_freq[mode];
+		for (dec = ARRAY_SIZE(ad7768_dec_rate) - 1; dec >= 0;  dec--) {
```
When counting down, prefer to do `for (dec = ARRAY_SIZE(ad7768_dec_rate); dec > 0;  dec--)` and use `dec - 1` in place of `dec` in the loop. This permits it to be typed `unsigned int`.

pcercuei commented on drivers/iio/adc/cf_axi_adc.h:
```diff
@@ -43,6 +44,9 @@
 #define ADI_MANUAL_SYNC_REQUEST		(1 << 8)
 
+#define ADI_REG_CNTRL_3			0x004c
+#define ADI_CRC_EN(x)			(((x) & 0x1) << 8)
```
This can just be `#define ADI_CRC_EN    BIT(8)`

nunojsa replied:
Indeed, but I think (didn't really looked) it's just following the style of the header. So, I would say it needs a different patch to fix the style altogether... Buuut, I'm not sure if it's worth the trouble since I want to deprecate this thing :smile:

pcercuei replied:
It doesn't follow the style of the header, if you look right above, 1-bit values are declared as argument-less macros. So at the very least, it should be:
`#define ADI_CRC_EN    (1 << 8)`

nunojsa replied:
Ohh right, it looks like this wants to be a more "generic" macro so you can enable/disable... But I would assume that default state is disabled and so we only care about enabling it.

pcercuei commented:
Overall in a pretty good shape, just a few comments above.

acusco commented:
V4 -> V5

- resolved nits
- rename dt-nodes to comply with DTSpec

nunojsa commented on drivers/iio/adc/ad7768.c:
```diff
@@ -313,7 +345,8 @@ static int ad7768_get_power_mode(struct iio_dev *dev,
 
-	st->power_mode = AD7768_POWER_MODE_GET_POWER_MODE(regval);
+	st->power_mode = ad7768_map_regval_to_power_mode(
+			 AD7768_POWER_MODE_GET_POWER_MODE(regval));
```
nit: I would refactor this to make checkpatch happy. This does not look that nice.... Maybe use a variable to store ` AD7768_POWER_MODE_GET_POWER_MODE(regval)` to make the line smaller...

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -41,6 +41,8 @@
 #define AD7768_POWER_MODE_MCLK_DIV_MSK		GENMASK(1, 0)
 #define AD7768_POWER_MODE_MCLK_DIV_MODE(x)	(((x) & 0x3) << 0)
+#define ad7768_map_power_mode_to_regval(x)	((x) ? x + 1 : 0)
+#define ad7768_map_regval_to_power_mode(x)	((x) ? x - 1 : 0)
```
Nit: about these two macros. You should protect `x + 1` with parentheses as well: `(x) + 1`.

pcercuei commented on drivers/iio/adc/ad7768.c:
```diff
@@ -52,25 +54,36 @@
 	AD7768_LOW_POWER_MODE,
-	AD7768_MEDIAN_MODE = 2,
-	AD7768_FAST_MODE
+	AD7768_MEDIAN_MODE,
```
`AD7768_MEDIAN_MODE` changed from 2 to 1, is that normal?

acusco replied:
Yes, because I made the logic much simpler in this case and just defined the mappings to and from register values to power modes and keep the enum values contiguous to easy work with the arrays of available sampling frequencies for each power mode

pcercuei reviewed (approved):
LGTM, with the two concerns above.

nunojsa reviewed (approved):
It looks like the dt schema error is due to some issue in CI
